test/network/network_test.cpp

```diff
@@ -0,0 +1,366 @@
+    }
+    LOG_DEBUG("Server initialized");
+    server_thread = std::thread([&]() { server.ServerLoop(); });
+  }
+
+  void TearDown() override {
+    server.Close();
+    server_thread.join();
+    LOG_DEBUG("Terrier has shut down");
+
+    TerrierTest::TearDown();
+  }
+};
+
+/**
+ * Use std::thread to initiate peloton server and pqxx client in separate
+ * threads
+ * Simple query test to guarantee both sides run correctly
+ * Callback method to close server after client finishes
+ */
+// NOLINTNEXTLINE
+TEST_F(NetworkTests, SimpleQueryTest) {
+  try {
+    pqxx::connection C(
+        fmt::format("host=127.0.0.1 port={0} user=postgres sslmode=disable application_name=psql", port));
+
+    pqxx::work txn1(C);
+    txn1.exec("INSERT INTO employee VALUES (1, 'Han LI');");
+    txn1.exec("INSERT INTO employee VALUES (2, 'Shaokun ZOU');");
+    txn1.exec("INSERT INTO employee VALUES (3, 'Yilei CHU');");
+
+    pqxx::result R = txn1.exec("SELECT name FROM employee where id=1;");
+    txn1.commit();
+    EXPECT_EQ(R.size(), 0);
+  } catch (const std::exception &e) {
+    LOG_ERROR("[SimpleQueryTest] Exception occurred: {0}", e.what());
+    EXPECT_TRUE(false);
+  }
+  LOG_DEBUG("[SimpleQueryTest] Client has closed");
+}
+
+ssize_t ReadUntilReadyOrClose(char *in_buffer, size_t max_len, int socket_fd) {
+  ssize_t n;
+  while (true) {
+    n = read(socket_fd, in_buffer, max_len);
+    if (n == 0 || in_buffer[n - 6] == 'Z')  // Ready for request
+      break;
+  }
+  return n;
+}
+
+/* strlcpy based on OpenBSDs strlcpy.
+ * this is a safer version of strcpy.
+ * clang-tidy does not accept strcpy so we need this function.
+ *
+ * Copy src to string dst of size siz.  At most siz-1 characters
+ * will be copied.  Always NUL terminates (unless siz == 0).
+ * Returns strlen(src); if retval >= siz, truncation occurred.
+ */
+size_t strlcpy(char *dst, const char *src, size_t siz) {
+  char *d = dst;
+  const char *s = src;
+  size_t n = siz;
+
+  /* Copy as many bytes as will fit */
+  if (n != 0 && --n != 0) {
+    do {
+      if ((*d++ = *s++) == 0) break;
+    } while (--n != 0);
+  }
+
+  /* Not enough room in dst, add NUL and traverse rest of src */
+  if (n == 0) {
+    if (siz != 0) *d = '\0'; /* NUL-terminate dst */
+    while ((*s++) != 0) {
+    }
+  }
+
+  return (s - src - 1); /* count does not include NUL */
+}
+
+int StartConnection(uint16_t port) {
+  // Manually open a socket
+  int socket_fd = socket(AF_INET, SOCK_STREAM, 0);
+
+  struct sockaddr_in serv_addr;
+  memset(&serv_addr, 0, sizeof(serv_addr));
+  serv_addr.sin_family = AF_INET;
+  serv_addr.sin_addr.s_addr = inet_addr("127.0.0.1");
+  serv_addr.sin_port = htons(port);
+
+  int64_t ret = connect(socket_fd, reinterpret_cast<sockaddr *>(&serv_addr), sizeof(serv_addr));
+  if (ret < 0) LOG_ERROR("Connection Error");
+
+  // Build the startup message
+  char out_buffer[TEST_BUF_SIZE] = {};
+  char in_buffer[TEST_BUF_SIZE] = {};
+  // 3: protocol version number
+  out_buffer[5] = 3;
```

Comment:
5?
This should be done differently. Actual implementation can be done several ways, but the general approach should be one of defining the packet structure and then using the fields to get and set values.

Peloton defines in marshal.h, struct OutputPacket for instance. 
Terrier defines in storage/projected_row.h, a class, used for MEM_REINTERPRETATION_ONLY, overlaying structure onto a byte buffer.

Input and output packets should use an approach as above, or equivalent. This comment also should be taken to apply to any other place where packet contents are read and written, in the code or tests.

Reply:
@DarkForte

Reply:
Later often never happens :-)

Reply:
Sorry for not keeping you updated. T_T
I was trying to support it in PostgresPacketWriter by adding more functions since Traffic Cop needs a utility to send the ResultSet back to the client anyway. I think we can just tell PostgresPacketWriter everything and it will write the packet for us. If that succeeded, all the packet writing code can be eliminated.